lib/utils/typecast_utils.dart

```diff
@@ -36,4 +37,65 @@ class TypeCast {
 
     return null;
   }
+
+  /// Checks for nullable type casts
+  /// Only one case `Type? is Type` always valid assertion case.
+  bool isNullableCompatibility() {
+    final isObjectTypeNullable =
+        source.nullabilitySuffix != NullabilitySuffix.none;
+    final isCastedTypeNullable =
+        target.nullabilitySuffix != NullabilitySuffix.none;
+
+    return isObjectTypeNullable && !isCastedTypeNullable;
+  }
+
+  /// Checks that type checking is unnecessary
+  /// [source] is the source expression type
+  /// [target] is the type against which the expression type is compared
+  /// [isReversed] true for opposite comparison, i.e 'is!'
+  /// and false for positive comparison, i.e. 'is', 'as' or 'whereType'
+  bool isUnnecessaryTypeCheck({
+        bool isReversed = false,
+      }) {
+    if (isNullableCompatibility()) {
+      return false;
+    }
+
+    final objectCastedType = castTypeInHierarchy();
+
+    if (objectCastedType == null) {
+      return isReversed;
+    }
+
+    final objectTypeCast = TypeCast(
+      source: objectCastedType,
+      target: target,
+    );
+
+    if (!objectTypeCast.areGenericsWithSameTypeArgs()) {
+      return false;
+    }
+
+    return !isReversed;
+  }
```

Comment:
I like the spacing, but when it's after every single statement -- it begins to lose meaning. Let's group minimally.
```suggestion
  bool isUnnecessaryTypeCheck({
        bool isReversed = false,
      }) {
    if (isNullableCompatibility()) return false;

    final objectCastedType = castTypeInHierarchy();
    if (objectCastedType == null) {
      return isReversed;
    }

    final objectTypeCast = TypeCast(
      source: objectCastedType,
      target: target,
    );
    if (!objectTypeCast.areGenericsWithSameTypeArgs()) {
      return false;
    }

    return !isReversed;
  }
```